src/compiler/types/validate-primary-package-output-target.ts

```diff
@@ -0,0 +1,222 @@
+import { buildWarn, isPrimaryPackageOutputTarget, isString, normalizePath } from '@utils';
+import { join, relative } from 'path';
+
+import type * as d from '../../declarations';
+import { packageJsonError, packageJsonWarn } from './package-json-log.utils';
+
+export type PrimaryPackageOutputTargetRecommendedConfig = {
+  getModulePath?: (rootDir: string, outputTargetDir: string) => string | null;
+  getTypesPath?: (rootDir: string, outputTargetConfig: any) => string | null;
+};
+
+export const PRIMARY_PACKAGE_TARGET_CONFIGS = {
+  dist: {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDist) =>
+      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
+  },
+  'dist-collection': {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+  },
+  'dist-custom-elements': {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistCustomElements) => {
+      return outputTargetConfig.generateTypeDeclarations
+        ? normalizePath(relative(rootDir, join(rootDir, outputTargetConfig.dir, 'index.d.ts')))
+        : null;
+    },
+  },
+  'dist-types': {
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistTypes) =>
+      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
+  },
+} satisfies Record<d.PrimaryPackageOutputTarget['type'], PrimaryPackageOutputTargetRecommendedConfig>;
+
+/**
+ * Performs validation for specified fields in a Stencil project's
+ * `package.json` based on output targets being designated as
+ * `isPrimaryPackageOutputTarget`.
+ *
+ * @param config The Stencil project's config.
+ * @param compilerCtx The project's compiler context.
+ * @param buildCtx The project's build context.
+ */
+export const validatePrimaryPackageOutputTarget = (
+  config: d.ValidatedConfig,
+  compilerCtx: d.CompilerCtx,
+  buildCtx: d.BuildCtx
+) => {
+  // TODO(NOW): should _all_ validation be gated behind the config flag, or just this
+  // new layer of "primary" target validation?
```

Comment:
I think this is the only outstanding question. The current implementation seems like the right approach, but we can gate _all_ validation behind the config flag if we wanted.

Also, at the time of writing this, I'm realizing I didn't leave the previous validation logic in place if the new config flag wasn't enabled. Do we want to have the old logic run if the new flag isn't enabled?